main/src/ca/uwaterloo/flix/language/ast/SyntaxTree.scala

```diff
@@ -0,0 +1,489 @@
+     * A [[SyntaxTree]] child holding a [[Token]].
+     */
+    case class NodeToken(token: Token) extends Child
+
+    /**
+     * A [[SyntaxTree]] child holding a nested [[SyntaxTree.Tree]]
+     */
+    case class NodeTree(tree: SyntaxTree.Tree) extends Child
+  }
+
+
+  /**
+   * A common super-type for [[TreeKind]]s
+   */
+  sealed trait TreeKind
+
+  /**
+   * Different kinds of syntax nodes in a [[SyntaxTree]].
+   * The only error kind that holds data is the special [[TreeKind.ErrorTree]].
+   */
+  object TreeKind {
+    /**
+     * A special error kind wrapping a [[Parser2Error]].
+     */
+    case class ErrorTree(error: Parser2Error) extends TreeKind
+
+    case object AnnotationList extends TreeKind
+
+    case object Argument extends TreeKind
+
+    case object ArgumentNamed extends TreeKind
+
+    case object ArgumentList extends TreeKind
+
+    case object Case extends TreeKind
+
+    case object CommentList extends TreeKind
+
+    case object Doc extends TreeKind
+
+    case object Ident extends TreeKind
+
+    case object Kind extends TreeKind
+
+    case object ModifierList extends TreeKind
+
+    case object Operator extends TreeKind
+
+    case object Parameter extends TreeKind
+
+    case object ParameterList extends TreeKind
+
+    case object QName extends TreeKind
+
+    case object Source extends TreeKind
+
+    case object TypeParameter extends TreeKind
+
+    case object TypeParameterList extends TreeKind
+
+    //////////////////////////////////////////////////////////////////////////////////////////
+    /// DECLARATIONS /////////////////////////////////////////////////////////////////////////
+    //////////////////////////////////////////////////////////////////////////////////////////
+    sealed trait Decl extends TreeKind
+
+    object Decl {
+      case object AssociatedTypeDef extends Decl
+
+      case object AssociatedTypeSig extends Decl
+
+      case object Class extends Decl
+
+      case object Def extends Decl
+
+      case object Effect extends Decl
+
+      case object Enum extends Decl
+
+      case object Instance extends Decl
+
+      case object Law extends Decl
+
+      case object Module extends Decl
+
+      case object Operation extends Decl
+
+      case object RestrictableEnum extends Decl
+
+      case object Signature extends Decl
+
+      case object TypeAlias extends Decl
+    }
+
+    //////////////////////////////////////////////////////////////////////////////////////////
+    /// EXPRESSIONS //////////////////////////////////////////////////////////////////////////
+    //////////////////////////////////////////////////////////////////////////////////////////
+    sealed trait Expr extends TreeKind
+
+    object Expr {
+
+      /**
+       * A marker kind used to wrap nested expressions.
+       * For instance on a binary expression "1 + 2" you would do
+       * Expr
+       * Binary
+       * Expr
+       * LiteralNumber
+       * Operator
+       * Expr
+       * LiteralNumber
```

Comment:
Formatting lost? :)

Reply:
dang

Reply:
Might have to loose those comments, I wasn't too sure about them either

Reply:
I like the comments, when they had formatting. I  think if you use // the formatter will not destroy them.